lab/lab-guide.md

```diff
@@ -0,0 +1,802 @@
+ 
+    ```csharp
+    public async Task<List<Product>> SearchProductsAsync(float[] vectors, int productMaxResults)
+    {
+        var options = new VectorSearchOptions { VectorPropertyName = "vectors", Top = productMaxResults };
+        #pragma warning disable SKEXP0020
+        var searchResult = await _productContainer.VectorizedSearchAsync(promptVectors, options);
+        #pragma warning restore SKEXP0020
+
+        var resultRecords = new List<VectorSearchResult<Product>>();
+        await foreach (var result in searchResult.Results)
+        {
+            resultRecords.Add(result);
+        }
+
+        //Serialize List<Product> to a JSON string to send to OpenAI
+        string productsString = JsonSerializer.Serialize(resultRecords);
+        return productsString;
+    }
+    ```
+
+If you get other strange behavior for the completion, it's possible the system prompt is not correct. In the **SemanticKernelService** locate the system prompts at the top of the class. Review `_systemPromptRetailAssistant` variable to make sure that your code matches this sample.
+ 
+    ```csharp
+    private readonly string _systemPromptRetailAssistant = @"
+    You are an intelligent assistant for the Cosmic Works Bike Company. 
+    You are designed to provide helpful answers to user questions about 
+    bike products and accessories provided in JSON format below.
+
+    Instructions:
+    - Only answer questions related to the information provided below,
+    - Don't reference any product data not provided below.
+    - If you're unsure of an answer, you can say ""I don't know"" or ""I'm not sure"" and recommend users search themselves.
+
+    Text of relevant information:";
+    ```
+
+Finally, if the responses do not include any information on the bike products being asked, it's possible the call to OpenAI Service is not correct. In the **SemanticKernelService** locate the `GetRagCompletionAsync()` method to make sure that your code matches this sample.
+ 
+    ```csharp
+    public async Task<(string completion, int tokens)> GetRagCompletionAsync(string sessionId, List<Message> contextWindow, string promptText, int productMaxResults)
+    {
+        float[] promptVectors = await GetEmbeddingsAsync(promptText);
+        string productsString = await SearchProductsAsync(promptVectors, productMaxResults);
+
+        var skChatHistory = new ChatHistory();
+        skChatHistory.AddSystemMessage(_systemPromptRetailAssistant + productsString);
+        
+        foreach (var message in contextWindow)
+        {
+            skChatHistory.AddUserMessage(message.Prompt);
+            if (message.Completion != string.Empty)
+                skChatHistory.AddAssistantMessage(message.Completion);
+        }
+
+        PromptExecutionSettings settings = new()
+        {
+            ExtensionData = new Dictionary<string, object>()
+            {
+                { "Temperature", 0.2 },
```

Comment:
btw, I looked at the source for Semantic Kernel and changed the values for the keys in this Dictionary to match what it is expecting.